src/viam/sdk/services/mlmodel/client.cpp

```diff
@@ -0,0 +1,158 @@
+// Copyright 2023 Viam Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <viam/sdk/services/mlmodel/client.hpp>
+
+#include <viam/sdk/services/mlmodel/private/proto.hpp>
+
+#include <grpcpp/channel.h>
+
+namespace viam {
+namespace sdk {
+
+namespace {
+
+struct tensor_storage_and_views {
+    mlmodel_details::tensor_storage storage;
+    MLModelService::named_tensor_views views;
+};
+
+}  // namespace
+
+MLModelServiceClient::MLModelServiceClient(std::string name, std::shared_ptr<grpc::Channel> channel)
+    : MLModelServiceClient(std::move(name), service_type::NewStub(channel)) {
+    channel_ = std::move(channel);
+}
+
+MLModelServiceClient::MLModelServiceClient(std::string name,
+                                           std::unique_ptr<service_type::StubInterface> stub)
+    : MLModelService(name), stub_(std::move(stub)) {}
+
+std::shared_ptr<MLModelService::named_tensor_views> MLModelServiceClient::infer(
+    const named_tensor_views& inputs) {
+    namespace pb = ::google::protobuf;
+    namespace mlpb = ::viam::service::mlmodel::v1;
+
+    pb::Arena arena;
+    auto* const req = pb::Arena::CreateMessage<mlpb::InferRequest>(&arena);
+
+    req->set_name(this->name());
+
+    auto& mutable_input_data = *req->mutable_input_data();
+    auto& mutable_input_data_fields = *mutable_input_data.mutable_fields();
+
+    for (const auto& kv : inputs) {
+        auto ib = mutable_input_data_fields.emplace(kv.first);
+        pb::Value& value = ib.first->second;
+        mlmodel_details::tensor_to_pb_value(kv.second, &value);
+    }
+
+    auto* const resp = pb::Arena::CreateMessage<mlpb::InferResponse>(&arena);
+
+    grpc::ClientContext ctx;
+    auto result = stub_->Infer(&ctx, *req, resp);
+
+    // TODO: Metadata caching.
+    const auto md = metadata();
+
+    auto tsav = std::make_shared<tensor_storage_and_views>();
+    const auto& output_fields = resp->output_data().fields();
+    for (const auto& output : md.outputs) {
+        const auto where = output_fields.find(output.name);
+        // Ignore any outputs for which we don't have metadata, since
+        // we can't know what type they should decode to.
+        if (where != output_fields.end()) {
+            mlmodel_details::pb_value_to_tensor(
+                output, where->second, &tsav->storage, &tsav->views);
+        }
+    }
+
+    return {std::move(tsav), &tsav->views};
```

Comment:
Tricky: this shared pointer keeps `tsav` alive but returns a shared_ptr which refers to the tensor views that `tsav` manages.

I'm pretty sure this works but I intend to write some tests to make sure.